### PR TITLE
Add ICryptoBackend.Sign

### PR DESCRIPTION
**Pull request by earlbread**
This adds `ICryptoBackend.Sign()` method to enable signing library changes. Also, this modifies `PrivateKey` to store `PublicKey` instead of generating it every time.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -33,6 +33,8 @@ To be released.
     `Block<T>.Mine()` static method.  [[#666], [#917]]
  -  Added `options` optional parameter to `Swarm<T>()` constructor.
     [[#926]]
+ -  Added `Sign(HashDigest<SHA256>, PrivateKey)` method to `ICryptoBackend`
+    interface.  [[#931]]
```
Comment (dahlia) [inline CHANGES.md:37]: Just for consistency with existing changelogs.

```suggestion
 -  Added `ICryptoBackend.Sign(HashDigest<SHA256>, PrivateKey)` method.
    [[#931]]
```

File: Libplanet/Crypto/ICryptoBackend.cs [hunk 2]
```diff
@@ -7,6 +7,20 @@ namespace Libplanet.Crypto
...
+        /// </summary>
+        /// <param name="messageHash">A 32 bytes message hash digest hashed with SHA256 to sign.
+        /// </param>
+        /// <param name="privateKey"><see cref="PrivateKey"/> to sign
+        /// <paramref name="messageHash"/>.
+        /// </param>
+        /// <returns> Created a signature from <paramref name="messageHash"/> with the corresponding
+        /// <paramref name="privateKey"/>.
+        /// </returns>
+        byte[] Sign(HashDigest<SHA256> messageHash, PrivateKey privateKey);
```
Comment (dahlia) [inline Libplanet/Crypto/ICryptoBackend.cs:22]: Why does this take `HashDigest<SHA256>` instead of any arbitrary `byte[]`?

Comment (earlbread) [reply Libplanet/Crypto/ICryptoBackend.cs:22]: Just for consistency with `ICryptoBackend.Verify()`. I don't remember the exact reason why `Verify` take `HashDigest<SHA256>`, but the discussion is here https://github.com/planetarium/libplanet/pull/758#discussion_r366175486. Would it be better `byte[]`?

Comment (dahlia) [reply Libplanet/Crypto/ICryptoBackend.cs:22]: Pair enough.  How about generalizing it over `SHA256`, e.g., `HashDigest<T>`?

Comment (earlbread) [reply Libplanet/Crypto/ICryptoBackend.cs]: Changed `ICryptoBackend` to `ICryptoBackend<T>`. 1a1e4d9

File: CHANGES.md [hunk 3]
```diff
@@ -33,6 +33,11 @@ To be released.
     `Block<T>.Mine()` static method.  [[#666], [#917]]
  -  Added `options` optional parameter to `Swarm<T>()` constructor.
     [[#926]]
+ -  `ICryptoBackend` became to `ICryptoBackend<T>`.  [[#932]]
+ -  `ICryptoBackend.Verify(HashDigest<SHA256>, byte[], PublicKey)` became to.
```
Comment (dahlia) [inline CHANGES.md:37]:
```suggestion
 -  `ICryptoBackend.Verify(HashDigest<SHA256>, byte[], PublicKey)` became to
```

File: Libplanet/Crypto/DefaultCryptoBackend.cs [hunk 4]
```diff
@@ -1,14 +1,41 @@
...
 using System.Security.Cryptography;
 using Org.BouncyCastle.Asn1;
+using Org.BouncyCastle.Crypto.Digests;
 using Org.BouncyCastle.Crypto.Signers;
+using Org.BouncyCastle.Math;
 
 namespace Libplanet.Crypto
 {
-    public class DefaultCryptoBackend : ICryptoBackend
+    public class DefaultCryptoBackend<T> : ICryptoBackend<T>
```
Comment (dahlia) [inline Libplanet/Crypto/DefaultCryptoBackend.cs:10]: This needs to be recorded to the changelog.

Comment (earlbread) [reply Libplanet/Crypto/DefaultCryptoBackend.cs:10]: I fixed the changelog and amended the commit. d19ca5b

File: Libplanet/Crypto/ICryptoBackend.cs [hunk 5]
```diff
@@ -5,8 +5,25 @@ namespace Libplanet.Crypto
     /// <summary>
     /// Cryptography backend interface.
     /// </summary>
-    public interface ICryptoBackend
+    /// <typeparam name="T">A <see cref="HashAlgorithm"/> which corresponds toa digest.</typeparam>
```
Comment (moreal) [inline Libplanet/Crypto/ICryptoBackend.cs:8]: I just wonder `toa` is a typo mistake or a word I didn't know.

Comment (earlbread) [reply Libplanet/Crypto/ICryptoBackend.cs:8]: It's a typo 😢 . Fixed it.

**Pull request merged**